fms/models/granite_v4.py

```diff
@@ -0,0 +1,623 @@
+    replacements = [
+        (r"^lm_head.weight", "head.weight"),
+        (r"^model.embed_tokens.weight", "base_model.embedding.weight"),
+        (r"^model.norm", "base_model.dec_norm"),
+        (r"^model.layers", "base_model.layers"),
+        (r"self_attn\.k_proj", "attn.in_proj.key"),
+        (r"self_attn\.v_proj", "attn.in_proj.value"),
+        (r"self_attn\.q_proj", "attn.in_proj.query"),
+        (r"self_attn\.o_proj", "attn.dense"),
+        (r"shared_mlp\.input_linear", "ff_sub_layer.wg1_fused"),
+        (r"shared_mlp\.output_linear", "ff_sub_layer.w2"),
+        (r"input_layernorm", "ln"),
+        (r"post_attention_layernorm", "ff_ln"),
+    ]
+    new_sd = {}
+    for name, param in input_sd.items():
+        new_name = name
+        for pattern, repl in replacements:
+            new_name = re.sub(pattern, repl, new_name)
+        new_sd[new_name] = param
+    return new_sd
+
+
+serialization.register_adapter_step(
+    _architecture_name, "hf_to_fms_names", _hf_to_fms_names
+)
+
+
+serialization.register_adapter_step(
+    _architecture_name,
+    "weight_expansion_for_mismatched_head_dim",
+    serialization._weight_expansion_for_mismatched_head_dim,  # type: ignore[arg-type]
+)
+
+
+def _get_rope_params(linear_type: str) -> list[str]:
+    if "gptq" in linear_type:
+        return ["qweight", "scales", "qzeros", "bias"]
+    elif "int8" in linear_type:
+        # quantize_weight is fms-model-optimizer identifier of weight clip values
+        return ["weight", "bias", "quantize_weight"]
+    elif "fp8" in linear_type:
+        return ["weight", "weight_scale", "input_scale", "bias"]
+    else:  # torch.nn.Linear
+        return ["weight", "bias"]
+
+
+def _hf_to_fms_rope(
+    input_sd: Mapping[str, Any],
+    model_config: Optional[GraniteMoeHybridConfig] = None,
+    **kwargs,
+) -> Mapping[str, Any]:
+    new_sd = {}
+
+    if model_config:
+        head_size = model_config.emb_dim // model_config.nheads
+    else:
+        logger.warning("Missing model_config, assuming defaults for head_size")
+        head_size = 128  # Good default for most models
+
+    for name, param in input_sd.items():
+        # Some checkpoints have weights in different precisions, which can have
+        # auxiliary tensors (see _get_rope_params e.g. gptq, fp8).
+        # Thus, we need to get rope_params per parameter.
+        linear_type_str = "torch_linear"
+        if model_config and model_config.linear_config:
+            linear_type_str = get_linear_type(
+                model_config.linear_config,
+                module_name=name,
+            )
+        rope_params = _get_rope_params(linear_type_str)
+        trans_required_pattern = re.compile(
+            f"base_model.layers.[0-9]+.attn.in_proj.(query|key).({'|'.join(rope_params)})$"
+        )
+
+        # hf -> fms requires a transpose operation for the query and key
+        # weight and bias parameters for Llama models
```

Comment:
Since this is now applied to Granite 4 as well, it might be clearer to make that comment architecture-agnostic or explicitly say “Llama/Granite”.